src/useFormState.js

```diff
@@ -81,20 +84,49 @@ export default function useFormState(initialState, options) {
     function validate(e, values = formState.current.values) {
       let error;
       let isValid = true;
-      if (isFunction(inputOptions.validate)) {
-        const result = inputOptions.validate(e.target.value, values, e);
+
+      const value = isString(e) ? e : e.target.value;
```

Comment:
I would maybe flip this around and do:

```suggestion
      const value = e && e.target ? e.target.value : e;
```

There's a couple of cases I can think of where `e` might not be a string; `react-select`, for example, will let you pick multiple values from a select, in which case you'd expect it's onChange to give you a `string[]`, but it actually it gives you back a `{value: string, label: string} | {value: string, label: string}[]`.  It expects the same in it's `.value` too, which is kind of obnoxious, but that's the way it is.  :P

In our product we have a custom phone number input component where the `onChange` and `value` are a `{country: string, number: string}`, too.

If you change this, then a similar comment applies to the customValidate logic below.

Reply:
Related; if we're not expecting a string, the type of `value` will be wrong [here](https://github.com/wsmd/react-use-form-state/blob/d44bdaa7b357aa81144e7a61905c701491036c31/src/index.d.ts#L122) for us typescript folk.  You *could* fix this by making it `value: any` here, although it's obviously not type safe.

Reply:
Hmmmm I see where this is going...

I was trying to approach this as if the value is going to be a string, which is essentially the value of all input types except for checkboxes (which can be an array or a boolean).

I think we're going down the rabbit hole here. This will impose limitations moving forward – it's starting to not make sense anymore – with this approach inputs can lose their semantics – an input of type text can literally be anything.. it's no longer safe to assume that input values will be strings....

You know what.. I think I see the value with `raw` – it's safe to assume that whenever something `isRaw` that value will be anything. This immediately clicked as soon as I saw this line:

https://github.com/wsmd/react-use-form-state/blob/bf04f7c3f52aa433951b5ed3efafe7a705077100/src/useFormState.js#L149-L152

I'd like to give `.raw` another chance – you've definitely thought this through, but I'd like to iterate on a couple of things – let's move the conversation over there – I'm sure we can find a middle ground between the two approaches 😄